src/readers/morphologyHDF5.cpp

```diff
@@ -289,8 +288,17 @@ int MorphologyHDF5::_readV1Sections() {
     //            forcing HDF5 + MPI-IO to read in 4-byte groups. Thus, we now
     //            read the whole dataset at once, and split it in memory.
 
-    std::vector<std::vector<int>> vec;
-    _sections->read(vec); 
+    const auto dims = _sections->getSpace().getDimensions();
+
+    if (dims.size() != 2 || dims[1] != _structureV1Columns) {
+        throw(MorphioError("Error reading morphologies " + _uri +
+                           " bad number of dimensions in 'structure' dataspace"));
+    }
+
+    std::vector<std::array<int, _structureV1Columns>> vec(dims[0]);
+    if (vec.size() > 0) {
+        _sections->read(vec.front().data());
```

Comment:
This strikes me as dangerous. `data()` returns a bare pointer; are we sure the allocated size will be enough no matter what? Because `_sections->read()` can't do any bounds-checking (which I assume it did before?)

Reply:
I don't like it either, but it's the same logic essentially, just shifted from HighFive to here and avoiding a bunch of inner vector allocations. There's a dimension check before this…

Reply:
The HighFive logic is evolving, and should become better over time, no? By doing it ourselves, we have to backport that.

Can't HighFive do a `vec.reserve()` to avoid the inner vec alloations?

> There's a dimension check before this…

Are type sizes considered (along w/ the widening narrowing)?

Reply:
HighFive should become better, yes. But that may still be off for a while.

> Can't HighFive do a `vec.reserve()` to avoid the inner vec alloations?

On the outer vector yes, but that leaves individual allocations for thousands of inner vectors.

Reply:
> On the outer vector yes, but that leaves individual allocations for thousands of inner vectors.

Ah, it can't use std::array like is done here?

Reply:
Not right now.

Reply:
FWIW, I _think_ this is the logic to get HDF5 to read with the right datatype, even if we just pass in a ptr:

https://github.com/BlueBrain/HighFive/blob/4c70d818ed18231563fe49ff197d1c41054be592/include/highfive/bits/H5Slice_traits_misc.hpp#L183

Reply:
I think there should be comment that reminds us to go back to the (safer, imo) way once HighFive does multi-dim with std::array.